### PR TITLE
Merge the configured and the running broker ids when reconciling services

### PR DESCRIPTION
**Pull request by adamantal**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| API breaks?     | no
| Deprecations?   | no
| Related tickets | fixes regression from #457 
| License         | Apache 2.0

### What's in this PR?

Fixes a regression introduced by #457.

Tested locally in docker-desktop.
After specifying external listener config in cold start the reconcile succeeds.

### Why?

See more information at #440.

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -179,14 +179,28 @@ func AreStringSlicesIdentical(a, b []string) bool {
 	return reflect.DeepEqual(a, b)
 }
 
-func GetBrokerIdsFromStatus(brokerStatuses map[string]v1beta1.BrokerState, log logr.Logger) []int {
-	brokerIds := make([]int, 0, len(brokerStatuses))
+// returns the union of the ids of the configured (Spec.Brokers) and the running (BrokerState) brokers
+func GetBrokerIdsFromStatusAndSpec(brokerStatuses map[string]v1beta1.BrokerState, brokers []v1beta1.Broker, log logr.Logger) []int {
+	brokerIdMap := make(map[int]bool)
```
Comment (stoader) [inline pkg/util/util.go:184]:
```suggestion
	brokerIdMap := make(map[int]struct{})
```

File: pkg/util/util.go [hunk 2]
```diff
@@ -179,14 +179,28 @@ func AreStringSlicesIdentical(a, b []string) bool {
...
+
+	// add brokers from spec
+	for _, broker := range brokers {
+		brokerIdMap[int(broker.Id)] = true
```
Comment (stoader) [inline pkg/util/util.go:188]:
```suggestion
		brokerIdMap[int(broker.Id)] = struct{}{}
```

File: pkg/util/util.go [hunk 3]
```diff
@@ -179,14 +179,28 @@ func AreStringSlicesIdentical(a, b []string) bool {
...
+	}
+
+	// add brokers from status
 	for brokerId := range brokerStatuses {
 		id, err := strconv.Atoi(brokerId)
 		if err != nil {
 			log.Error(err, "could not parse brokerId properly")
 			continue
 		}
+		brokerIdMap[id] = true
```
Comment (stoader) [inline pkg/util/util.go:198]:
```suggestion
		brokerIdMap[id] = struct{}{}
```

File: pkg/util/util.go [hunk 4]
```diff
@@ -179,14 +179,28 @@ func AreStringSlicesIdentical(a, b []string) bool {
...
+	}
+
+	// collect unique broker ids
+	brokerIds := make([]int, 0)
```
Comment (stoader) [inline pkg/util/util.go:202]:
```suggestion
	brokerIds := make([]int, 0, len(brokerIdMap))
```

**Review by stoader — CHANGES_REQUESTED** [review]

**Comment by amuraru** [comment]
thanks @adamantal for tagging me on https://github.com/banzaicloud/kafka-operator/issues/440#issuecomment-729771482
I am curious is that ingress reconcile succeeds eventually on a fresh start after the first kafka pod is created and a status entry is added?

**Comment by adamantal** [comment]
> thanks @adamantal for tagging me on [#440 (comment)](https://github.com/banzaicloud/kafka-operator/issues/440#issuecomment-729771482)
> I am curious is that ingress reconcile succeeds eventually on a fresh start after the first kafka pod is created and a status entry is added?

Unfortunately no, the reconcile does not succeed eventually. 
The creation of broker pods happens on a later point in time, so we fail sooner with the error about the bad service spec.

However this does not happen if you have an already running cluster because there are valid broker statuses.

**Comment by adamantal** [comment]
Added unit tests in latest commit.

**Review by stoader — APPROVED** [review]

**Review by baluchicken — APPROVED** [review]

**Pull request merged**